enhancements/two-nodes-openshift/2no.md

```diff
@@ -0,0 +1,664 @@
+
+MCO - Machine Config Operator. This operator manages updates to the node's systemd, cri-o/kubelet, kernel, NetworkManager, etc., and can write custom files to it, configurable by MachineConfig custom resources.
+
+ABI - Agent-Based Installer.
+
+BMO - Baremetal Operator
+
+CEO - cluster-etcd-operator
+
+BMC - Baseboard Management Console. Used to manage baremetal machines. Can modify firmware settings and machine power state.
+
+## Summary
+
+Leverage traditional high-availability concepts and technologies to provide a container management solution that has a minimal footprint but remains resilient to single node-level failures suitable for customers with numerous geographically dispersed locations.
+
+## Motivation
+
+Customers with hundreds, or even tens of thousands, of geographically dispersed locations are asking for a container management solution that retains some level of resilience to node-level failures but does not come with a traditional three-node footprint and/or price tag.
+
+The need for some level of fault tolerance prevents the applicability of Single Node OpenShift (SNO), and a converged 3-node cluster is cost prohibitive at the scale of retail and telcos - even when the third node is a "cheap" one that doesn't run workloads.
+
+The benefits of the cloud-native approach to developing and deploying applications are increasingly being adopted in edge computing.
+This requires our solution to provide a management experience consistent with "normal" OpenShift deployments and be compatible with the full ecosystem of Red Hat and partner workloads designed for OpenShift.
+
+### User Stories
+
+* As a large enterprise with multiple remote sites, I want a cost-effective OpenShift cluster solution so that I can manage containers without the overhead of a third node.
+* As a support engineer, I want a safe and automated method for handling the failure of a single node so that the downtime of the control-plane is minimized.
+* As an enterprise running workloads on a minimal OpenShift footprint, I want to minimize time-to-recovery and data loss for my workloads when a node fails.
+
+### Goals
+
+* Provide a two-node control-plane for physical hardware that is resilient to a node-level failure for either node
+* Provide a transparent installation experience that starts with exactly 2 blank physical nodes, and ends with a fault-tolerant two-node cluster
+* Prevent both data corruption and divergent datasets in etcd
+* Minimize recovery-caused unavailability. Eg. by avoiding fencing loops, wherein each node powers cycles its peer after booting, reducing the cluster's availability.
+* Recover the API server in less than 120s, as measured by the surviving node's detection of a failure
+* Minimize any differences to existing OpenShift topologies
+* Avoid any decisions that would prevent future implementation and support for upgrade/downgrade paths between two-node and traditional architectures
```

Comment:
Yes - it's specifically calling out going from 2NO to HA-Compact.

This is because while support for that is a non-goal, PM has made it clear that there is a strong interest in this as a follow-on feature. Since we don't have requirements for this, we wanted to point out that while it's a secondary (or tertiary) goal, we'd like to avoid implementations that will create lots of tech debt if requirements firm up for us.

Reply:
Standalone compact being a three node control plane with no workers?

What `controlPlaneToplogy` is a three node compact? And does that mean we then need to explore more about the issues with operators that are not compatible today with changing toplogy?

Reply:
Yes, to standalone compact there are 3 control plane nodes and 0 compute.
The topology used for anything with 3 or more control plane nodes is `HighlyAvailable`

I think we need to explore topology transitions as their own enhancement proposal.